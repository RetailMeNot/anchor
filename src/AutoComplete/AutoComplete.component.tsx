// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled from 'styled-components';
// COMPONENTS
import { List, Item } from '../List';
import { Icon } from '../Icon';
// THEME
import { fonts, colors } from '../theme';

// TODO: This should be built from a FormControl Input, but we don't have one yet. =(

interface AutoCompleteProps {
  dataSource: any[];
  className?: string;
  size?: 'small' | 'large';
  placeholder?: string;
  onSearch?: () => void;
  onSelect?: () => void;
  // onFocus?: () => void;
  // onBlur?: () => void;
  children?: any;

  suffix?: React.ReactElement<Icon>; // TODO: consider buttons & other elements
  prefix?: React.ReactElement<Icon>; // TODO: consider buttons & other elements
}

interface AutoCompleteState {
  isFocused: boolean;
}

export const StyledAutoComplete = styled.div`
  position: relative;
  border: solid 1px ${colors.neutrals.ash.light};
  border-radius: .25rem;
  transition: border-color 500ms;
  display: flex;
  background: ${colors.neutrals.white.base};
  align-items: center;

  &:hover {
    border: solid thin ${colors.neutrals.ash.base};
  }

  &:active,
  &:focus {
    border: solid thin ${colors.neutrals.ash.dark};
  }

  .auto-complete-input {
    flex: 1;
    font-family: ${fonts.fontFamily};
    font-weight: 500;
    border: none;
    border-radius: .25rem;
    background-color: transparent;
    line-height: 1rem;
    overflow: hidden;

    color: ${colors.neutrals.charcoal.light};

    padding: ${({ size = 'small' }: AutoCompleteProps) => size === 'large' ? '0.75rem 1rem' : '0.5rem 0.5rem'};

    width: ${({ size = 'small' }: AutoCompleteProps) =>
      size === 'large' ? 'calc(100% - 2rem - 2px)' : 'calc(100% - 1rem - 2px)'};

    font-size: ${({ size = 'small' }: AutoCompleteProps) => size === 'large' ? '1rem' : '.75rem'};
  }

  .auto-complete-results {
    z-index: 2;
    position: absolute;
    top: ${({ size = 'small' }: AutoCompleteProps) => size === 'large' ? 'calc(2.5rem - 2px)' : 'calc(2rem - 2px)'};
    width: 100%;
    background-color: ${colors.neutrals.white.base};
    box-shadow: 0 8px 8px 0 rgba(0, 0, 0, 0.12);
    border-radius: 0 0 .25rem .25rem;
    border-top: solid 1px ${colors.neutrals.ash.light};
    transition: border-top-color 500ms;

    &:hover {
      border-top-color: ${colors.neutrals.ash.base};
    }
  }

  .auto-complete-prefix {
    z-index: 1;
    flex: 0 0 1.5rem;
  }
  /* TODO: suffix */
`;

export class AutoComplete extends React.Component<AutoCompleteProps> {
  static defaultProps = {
    placeholder: 'Search',
  };
  readonly state: AutoCompleteState = {
    isFocused: false,
  };

  render(): React.ReactElement<AutoComplete> {
    const {
      className,
      placeholder,
      children,
      suffix,
      prefix,
      ...props
    } = this.props;

    return (
      <StyledAutoComplete className={classNames(className)} {...props}>
        {prefix && React.cloneElement(prefix, {
          className: 'auto-complete-prefix',
          height: 16,
          width: 16,
          color: colors.neutrals.ash.dark,
        })}
        <input
          onFocus={() => this.setState({ ...this.state, isFocused: true })}
          onBlur={() => this.setState({ ...this.state, isFocused: false })}
          placeholder={placeholder}
          className="auto-complete-input"
          name="auto-complete"
        />
        {suffix && React.cloneElement(suffix, { className: 'auto-complete-suffix' })}
        {children}
        {this.state.isFocused && !children &&
          <List title="Recent Searches" className="auto-complete-results">
            {this.props.dataSource.map((item: any) => <Item key={item}>{item}</Item>)}
          </List>
        }
      </StyledAutoComplete>
    );
  }
}

export default AutoComplete;
