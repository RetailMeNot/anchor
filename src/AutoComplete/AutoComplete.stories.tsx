// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled from 'styled-components';
// COMPONENTS
import AutoComplete from './AutoComplete.component';
import {Icon} from '../Icon';
// THEME

const StyledStory = styled.div`
  padding: 2rem 5rem;
`;

storiesOf('Components/AutoComplete', module)
  .add('Default', () => (
      <StyledStory>
        <p>AutoComplete 1</p>
        <AutoComplete prefix={<Icon type="search"/>} dataSource={[1, 2, 3, 4]} />
        <p>AutoComplete 2</p>
        <AutoComplete
          size="large"
          placeholder="Stores, brands, categories, etc."
          dataSource={[1, 2, 3, 4]}
          prefix={<Icon type="search"/>}
        />
      </StyledStory>
  ));
