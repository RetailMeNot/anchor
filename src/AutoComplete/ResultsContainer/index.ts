export { ResultsContainer } from './ResultsContainer.component';
