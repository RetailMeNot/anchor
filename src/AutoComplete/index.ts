export { default as AutoComplete } from './AutoComplete.component';

