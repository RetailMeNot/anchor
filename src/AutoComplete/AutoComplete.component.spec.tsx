// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// COMPONENT
import AutoComplete from './AutoComplete.component';
// ENZYME
import { shallow, mount } from 'enzyme';

// TEST SETUP
const subject = <AutoComplete dataSource={[1, 2, 3, 4]} />;
const wrapper = mount(subject);
const component = shallow(subject);

describe('Component: AutoComplete', () => {
    it('should be defined', () => {
        /* tslint:disable */
        console.log(subject);
        console.log(wrapper.debug());
        console.log(component.debug());

        /* tslint:enable */
        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
