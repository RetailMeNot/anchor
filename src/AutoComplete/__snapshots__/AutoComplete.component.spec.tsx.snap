// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: AutoComplete should be defined 1`] = `
.c1 {
  display: block;
  position: relative;
  border: solid thin #D3D3D3;
  border-radius: 4px;
  cursor: text;
  box-sizing: border-box;
  min-width: 5rem;
  width: 100%;
  margin: 0;
  overflow: hidden;
  height: 3.125rem;
  padding: 0.25rem;
}

.c1::-webkit-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1::-moz-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1:-ms-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1::placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1.focus {
  border-color: #808080;
}

.c1 label {
  -webkit-transform-origin: left bottom;
  -ms-transform-origin: left bottom;
  transform-origin: left bottom;
  height: 1.4rem;
  -webkit-transform: translate(0,0.6rem) scale(1);
  -ms-transform: translate(0,0.6rem) scale(1);
  transform: translate(0,0.6rem) scale(1);
}

.c1 label.lift {
  -webkit-transform: translate(0,0) scale(1);
  -ms-transform: translate(0,0) scale(1);
  transform: translate(0,0) scale(1);
}

.c1 input {
  height: 1rem;
  font-size: 1rem;
}

.c3 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column-reverse;
  -ms-flex-flow: column-reverse;
  flex-flow: column-reverse;
  padding: 0 0.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  box-sizing: border-box;
  border: none;
  padding: 0;
  outline: none !important;
  touch-action: manipulation;
  -webkit-appearance: none;
  background-color: transparent;
  z-index: 1;
  color: #323232;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
}

.c4[type='number']::-webkit-inner-spin-button,
.c4[type='number']::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c4:placeholder-shown + label {
  cursor: text;
  max-width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.c0 {
  width: 100%;
  position: relative;
  border: solid thin transparent;
  border-color: #D3D3D3;
  border-radius: 4px;
  -webkit-transition: border-color 250ms;
  transition: border-color 250ms;
  background-color: #FFFFFF;
  color: #595959;
}

.c0:hover {
  border-color: #808080;
}

.c0:active,
.c0.focus,
.c0:focus {
  border-color: #808080;
}

.c0 .auto-complete-input {
  border: none;
}

<div
  className="anchor-auto-complete c0"
  color="neutrals.charcoal.light"
>
  <div
    className="anchor-input auto-complete-input c1"
    onClick={[Function]}
    size="lg"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <input
          aria-label="auto-complete"
          className="c4"
          name="auto-complete"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          placeholder="Search..."
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Component: AutoComplete should render with a shadow 1`] = `
.c1 {
  display: block;
  position: relative;
  border: solid thin #D3D3D3;
  border-radius: 4px;
  cursor: text;
  box-sizing: border-box;
  min-width: 5rem;
  width: 100%;
  margin: 0;
  overflow: hidden;
  height: 3.125rem;
  padding: 0.25rem;
}

.c1::-webkit-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1::-moz-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1:-ms-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1::placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1.focus {
  border-color: #808080;
}

.c1 label {
  -webkit-transform-origin: left bottom;
  -ms-transform-origin: left bottom;
  transform-origin: left bottom;
  height: 1.4rem;
  -webkit-transform: translate(0,0.6rem) scale(1);
  -ms-transform: translate(0,0.6rem) scale(1);
  transform: translate(0,0.6rem) scale(1);
}

.c1 label.lift {
  -webkit-transform: translate(0,0) scale(1);
  -ms-transform: translate(0,0) scale(1);
  transform: translate(0,0) scale(1);
}

.c1 input {
  height: 1rem;
  font-size: 1rem;
}

.c3 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column-reverse;
  -ms-flex-flow: column-reverse;
  flex-flow: column-reverse;
  padding: 0 0.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  box-sizing: border-box;
  border: none;
  padding: 0;
  outline: none !important;
  touch-action: manipulation;
  -webkit-appearance: none;
  background-color: transparent;
  z-index: 1;
  color: #323232;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
}

.c4[type='number']::-webkit-inner-spin-button,
.c4[type='number']::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c4:placeholder-shown + label {
  cursor: text;
  max-width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.c0 {
  width: 100%;
  position: relative;
  border: solid thin transparent;
  border-color: #D3D3D3;
  border-radius: 4px;
  -webkit-transition: border-color 250ms;
  transition: border-color 250ms;
  background-color: #FFFFFF;
  color: #595959;
  box-shadow: 0 0.5rem 0.75rem -0.375rem rgba(0,0,0,0.12);
}

.c0:hover {
  border-color: #808080;
}

.c0:active,
.c0.focus,
.c0:focus {
  border-color: #808080;
}

.c0 .auto-complete-input {
  border: none;
}

<div
  className="anchor-auto-complete c0"
  color="neutrals.charcoal.light"
>
  <div
    className="anchor-input auto-complete-input c1"
    onClick={[Function]}
    size="lg"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <input
          aria-label="auto-complete"
          className="c4"
          name="auto-complete"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Component: AutoComplete should render with a suffix and prefix 1`] = `
.c1 {
  display: block;
  position: relative;
  border: solid thin #D3D3D3;
  border-radius: 4px;
  cursor: text;
  box-sizing: border-box;
  min-width: 5rem;
  width: 100%;
  margin: 0;
  overflow: hidden;
  height: 3.125rem;
  padding: 0.25rem;
}

.c1::-webkit-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1::-moz-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1:-ms-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1::placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1.focus {
  border-color: #808080;
}

.c1 label {
  -webkit-transform-origin: left bottom;
  -ms-transform-origin: left bottom;
  transform-origin: left bottom;
  height: 1.4rem;
  -webkit-transform: translate(0,0.6rem) scale(1);
  -ms-transform: translate(0,0.6rem) scale(1);
  transform: translate(0,0.6rem) scale(1);
}

.c1 label.lift {
  -webkit-transform: translate(0,0) scale(1);
  -ms-transform: translate(0,0) scale(1);
  transform: translate(0,0) scale(1);
}

.c1 input {
  height: 1rem;
  font-size: 1rem;
}

.c3 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column-reverse;
  -ms-flex-flow: column-reverse;
  flex-flow: column-reverse;
  padding: 0 0.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  box-sizing: border-box;
  border: none;
  padding: 0;
  outline: none !important;
  touch-action: manipulation;
  -webkit-appearance: none;
  background-color: transparent;
  z-index: 1;
  color: #323232;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
}

.c4[type='number']::-webkit-inner-spin-button,
.c4[type='number']::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c4:placeholder-shown + label {
  cursor: text;
  max-width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.c0 {
  width: 100%;
  position: relative;
  border: solid thin transparent;
  border-color: #D3D3D3;
  border-radius: 4px;
  -webkit-transition: border-color 250ms;
  transition: border-color 250ms;
  background-color: #FFFFFF;
  color: #595959;
}

.c0:hover {
  border-color: #808080;
}

.c0:active,
.c0.focus,
.c0:focus {
  border-color: #808080;
}

.c0 .auto-complete-input {
  border: none;
}

<div
  className="anchor-auto-complete c0"
  color="neutrals.charcoal.light"
>
  <div
    className="anchor-input auto-complete-input c1"
    onClick={[Function]}
    size="lg"
  >
    <div
      className="c2"
    >
      <div
        className="input-prefix"
      >
        prefix
      </div>
      <div
        className="c3"
      >
        <input
          aria-label="auto-complete"
          className="c4"
          name="auto-complete"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="input-suffix"
      >
        suffix
      </div>
    </div>
  </div>
</div>
`;

exports[`Component: AutoComplete should render without a border 1`] = `
.c1 {
  display: block;
  position: relative;
  border: solid thin #D3D3D3;
  border-radius: 4px;
  cursor: text;
  box-sizing: border-box;
  min-width: 5rem;
  width: 100%;
  margin: 0;
  overflow: hidden;
  height: 3.125rem;
  padding: 0.25rem;
}

.c1::-webkit-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1::-moz-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1:-ms-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1::placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c1.focus {
  border-color: #808080;
}

.c1 label {
  -webkit-transform-origin: left bottom;
  -ms-transform-origin: left bottom;
  transform-origin: left bottom;
  height: 1.4rem;
  -webkit-transform: translate(0,0.6rem) scale(1);
  -ms-transform: translate(0,0.6rem) scale(1);
  transform: translate(0,0.6rem) scale(1);
}

.c1 label.lift {
  -webkit-transform: translate(0,0) scale(1);
  -ms-transform: translate(0,0) scale(1);
  transform: translate(0,0) scale(1);
}

.c1 input {
  height: 1rem;
  font-size: 1rem;
}

.c3 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column-reverse;
  -ms-flex-flow: column-reverse;
  flex-flow: column-reverse;
  padding: 0 0.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  box-sizing: border-box;
  border: none;
  padding: 0;
  outline: none !important;
  touch-action: manipulation;
  -webkit-appearance: none;
  background-color: transparent;
  z-index: 1;
  color: #323232;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
}

.c4[type='number']::-webkit-inner-spin-button,
.c4[type='number']::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c4:placeholder-shown + label {
  cursor: text;
  max-width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.c0 {
  width: 100%;
  position: relative;
  border: solid thin transparent;
  border-color: transparent;
  border-radius: 4px;
  -webkit-transition: border-color 250ms;
  transition: border-color 250ms;
  background-color: #FFFFFF;
  color: #595959;
}

.c0:hover {
  border-color: transparent;
}

.c0:active,
.c0.focus,
.c0:focus {
  border-color: transparent;
}

.c0 .auto-complete-input {
  border: none;
}

<div
  className="anchor-auto-complete c0"
  color="neutrals.charcoal.light"
>
  <div
    className="anchor-input auto-complete-input c1"
    onClick={[Function]}
    size="lg"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <input
          aria-label="auto-complete"
          className="c4"
          name="auto-complete"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
